helios-client/src/main/java/com/spotify/helios/auth/CrtAuthProvider.java

```diff
@@ -0,0 +1,165 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.helios.auth;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.AsyncFunction;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import com.spotify.crtauth.CrtAuthClient;
+import com.spotify.crtauth.agentsigner.AgentSigner;
+import com.spotify.crtauth.exceptions.CrtAuthException;
+import com.spotify.crtauth.signer.Signer;
+import com.spotify.helios.client.HeliosRequest;
+import com.spotify.helios.client.RequestDispatcher;
+import com.spotify.helios.client.Response;
+
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.concurrent.ExecutionException;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+
+public class CrtAuthProvider implements AuthProvider {
+
+  private static final Logger log = LoggerFactory.getLogger(AuthProviderSelector.class);
+
+  private static final String CRT_HEADER = "X-CHAP";
+  private static final String AUTH_URI = "https://helios/_auth";
+
+  private final RequestDispatcher dispatcher;
+  private final String username;
+  private final CrtAuthClient crtClient;
+  private volatile String token;
+  private final Object lock = new Object();
+
+  public CrtAuthProvider(final Context context,
+                         final String authServer) {
+    this.dispatcher = context.dispatcher();
+    this.username = context.user();
+    try {
+      this.crtClient = new CrtAuthClient(new AgentSigner(), authServer);
+    } catch (CrtAuthException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  @VisibleForTesting
+  CrtAuthProvider(final RequestDispatcher dispatcher,
+                  final String authServer,
+                  final String username,
+                  final Signer signer) {
+    this.dispatcher = dispatcher;
+    this.username = username;
+    this.crtClient = new CrtAuthClient(signer, authServer);
+  }
+
+  @Override
+  public String currentAuthorizationHeader() {
+    // TODO (dxia) Should we use a lock here? It doesn't really matter if the token isn't
+    // thread-safe. We should get valid tokens each time. The lock just prevents
+    // another thread from doing extra work to get an extra token when it could just wait.
+    if (token == null) {
+      synchronized (lock) {
+        try {
+          token = renewAuthorizationHeader(null).get();
```

Comment:
We probably need to tweak the `AuthProvider` interface to work well with the stored token, or rethink something else.

Currently, with `AuthProviderSelector`, the first request is *always* sent without authorization. Upon getting a 401 back it will select an `AuthProvider` and proceed to call `renewAuthorizationHeader()`. Now, we could tweak that to only call `renewAuthorizationHeader()` if the `currentAuthorizationHeader()` returns null on the selected provider - BUT, if we do that and the current token is expired the request will failed. Is there a way to tell if a token is expired on the client-side?